# Bioinformatics for Influenza Genome Reconstruction: A Hands-On Workshop
# Day 02: Intermediate Commands

## Lesson Objectives
* Understand the usage of the `grep` command to search for specific patterns or strings within files, particularly in genomic data files such as fastq.
* Practice counting occurrences of specific patterns or strings within files using `grep`.
* Explore the functionality of pipes (`|`) to chain commands together and transfer the output of one command as input to another.

## Search strings and text in files
The `grep` command scans a provided file to identify lines that contain matches to specified strings or words. `grep` searches can be combined with Regular Expressions (regex), a standardized search language. Regular expressions are concise expressions/instructions that detail the exact manner of how to search for some text in a file.

The command has the following pattern:
`grep '<pattern>' <file>`

There are two important facts to remember about how grep works by default:
* grep searches line by line in `<file>` for a match to the search pattern (`<pattern>`)
* any (!) lines that match the term are returned

### Finding a sequence motif
Using `grep`, we can locate occurrences of a specific sequence within a `fastq`. For instance, we can search for a designated sequence, such as AACACC. 

```bash
# Go to day01 directory

# Unzip ERR10453087.fastq.gz

# Search for the motif AACACC in the fastq file
grep "AACACC" ERR10453087.fastq
```

Instead of printing numerous lines, which may overwhelm the display due to the prevalence of this common motif, let's focus on counting the number of lines that contain a match.

```bash
# Count the number of lines that contain a match
grep -c "AACACC" ERR10453087.fastq
```

### Pipes 
The bash pipe, represented by `|`, transfers the output of one command to another, following the syntax [command1] | [command2]. This enables seamless chaining of commands in bash scripting, promoting efficiency and versatility in command execution.

For example, if the output of a `grep` search is not supposed to be printed to terminal but rather to a new window, we can pipe the output of `grep` into `less`. This chaining of commands avoids that the terminal is "spammed" with the output of the `grep` search.

```bash
# Pipe output of grep to less and observe the difference :)
grep "AACACC" ERR10453087.fastq | less
```

## Counting sequences
The `-c` option is particularly valuable for counting the number of sequences in a `fastq` file due to their organized structure. Each sequence in these files is expected to have a single header line, starting with `@` for `fastq` files. Thus, counting the number of `@` equals counting the number of sequences in a `fastq` file.

```bash
# Count the number of sequences in ERR10453087.fastq
grep -c "^@" ERR10453087.fastq
```

While `@` symbolizes both the sequence header and the base quality score, simply counting lines beginning with `@` risks overestimation by including quality score lines. However, if the sequences originate from sequence reads, their headers are likely to be similar, often featuring a machine identifier code. This can be verified by examining the first 20 lines of a `fastq` file, where you'll likely observe a consistent string at the start of each line, indicative of sequence machine identifier codes.

```bash
# Inspect the first 20 lines of ERR10453087.fastq, hint: use the parameter -n

# Count the number of sequences in ERR10453087.fastq with a more precise search pattern

```

It's important to recognize that this approach isn't foolproof. In datasets with millions of reads, there's a possibility of encountering a sequence with a quality score that begins with characters `@ERR104`, similar to the first few letters of the identifier code, all of which are valid quality score values. 

For absolute certainty, especially in critical applications, it's advisable to employ dedicated tools explicitly designed for parsing FASTQ files.
