# Bioinformatics for Influenza Genome Reconstruction: A Hands-On Workshop
# Day 04: Advanced tools - Variant calling

## Lesson Objectives
* Learn to utilize clair3 

In this lesson we will use a variant caller that is built specifically for long read data. Having seen the results of our mapping the reads to the Pan-Influenza reference, we saw that a lot of the resulting information includes a lot of redundant information. In this lesson we will boil this information down to its essentials.

### Set up the correct environment

```bash
# look at all conda environments available to you

# activate the "clair3" environment

# see that clair3 is available to you
run_clair3.sh --version
``` 
### Set up the reference
In order to use the Pan-Influenza reference we have to unpack and index it for `clair3`.

```bash
# copy the reference fasta to day04, or "here" if you are in day04
cp /home/mf1/nambi24/flu-workflow/refs/download/Pan_Influenza_refseq_genomes.fna.gz .

# unpack the reference file
gunzip Pan_Influenza_refseq_genomes.fna.gz

# index the reference file
samtools faidx Pan_Influenza_refseq_genomes.fna
``` 

### Variant Calling

The command we will use is `run_clair3.sh` and we will have to provide a great number of arguments, which are described in the following list.

- `bam_fn` the mapped sample
- `ref_fn` the reference file your sample was mapped to
- `threads` number of threads you want to use, depending on the number of threads your CPU can handle, this parameter can improve the runtime through parallelization
- `platform` sequencing technology used
- `model_path` here you should choose the model corresponding to the same Nanopore chemistry and flowcell version that was used for sequencing. These models are provided by the `clair3` software itself.
- `include_all_ctgs` we want to call variants for all contigs
- `no_phasing_for_fa` for polyploid samples you could turn on phasing
- `output` we will get quite a few output files so here you provide an output directory for `clair3`. The directory will be created by the tool itself, you just choose a name here.

```bash
# variant calling of s1
run_clair3.sh --bam_fn=s1.filtered.sorted.bam --ref_fn=Pan_Influenza_refseq_genomes.fna --threads=2 --platform="ont" --model_path=/home/mf1/miniconda3/envs/clair3/bin/models/r941_prom_sup_g5014 --include_all_ctgs --no_phasing_for_fa --output=s1_clair3
```

There is another way of calling it listed below. You can use that style to better view all paramters that you want to provide to clair3, you would only use that styl in a script though.

```bash
# variant calling of s1
run_clair3.sh \
  --bam_fn=s1.filtered.sorted.bam \
  --ref_fn=Pan_Influenza_refseq_genomes.fna \
  --threads=2
  --platform="ont" \
  --model_path=/home/mf1/miniconda3/envs/clair3/bin/models/r941_prom_sup_g5014 \
  --include_all_ctgs \
  --no_phasing_for_fa \
  --output=s1_clair3
```

