# Bioinformatics for Influenza Genome Reconstruction: A Hands-On Workshop
# Day 04: Pipelines - nextflow

## Lesson Objectives
* Learn to utilize a ready-made nextflow pipeline for the analysis of a set of toy samples.

In this lesson we will use `nextflow` and the `flu-workflow` pipeline to calculate the results that we got by our own efforts in the previous hands-ons. This pipeline generates even more results and it compares the samples that we put in and calculates a phylogeny. 

In this lesson we will not need to set up a conda environment. `nextflow` was installed in the conda `base` environment and even though the pipeline uses all the tools that we used before and needs even more conda environment than we set up, `nextflow` also takes care of setting up and activating those conda environments


### The workflow

```bash
# go to the workflow folder
cd /home/mf1/nambi24/flu-workflow

# notice the pipeline called "main.nf"
ls -l
```

### Utilize the workflow

We have to run the pipeline with the flag `-with-conda` to make sure that conda environments are used. The datafolder we provide after the `--rawfastq` argument.

```bash
# run the pipeline on all samples
nextflow run main.nf --rawfastq ../data -with-conda
```
