# Bioinformatics for Influenza Genome Reconstruction: A Hands-On Workshop
# Day 03: Advanced tools - abricate

## Lesson Objectives
* Learn to utilize abricate for typing of an influenza assembly

In this lesson we will use `abricate` for typing of our sequence. We have seen before that our sample `s1` is an influenza A sample. Now that we have an assembly we can easily find out its subtype. One tool that can do that is `abricate` and we will use it here.

### Set up the correct environment

```bash
# look at all conda environments available to you

# activate the "abricate" environment

# see that abricate is available to you
abricate --version
``` 

### Typing with abricate

`abricate` has a number of options. We have to tell it which database to use with the argument `--db` and where to find it with the argument `--datadir`. In the database we provide there are representative sequences for influenza types A and B. There exist also representative sequences for the subtypes of these types, e.g. for H1N1, H3N2, Yamagata, etc.. . With the argument `-minid` we can set a threshold where we demand that the assembly has that percentage identity to the representative sequences. With `-mincov` we demand that at least that percentage of the representative sequences is covered.

```bash
# typing with abricate
abricate --datadir /home/mf1/nambi24/flu-workflow/catalogues/INSAFLU --db db_influenza_typing_v8 -minid 70 -mincov 60 s1_assembly.fasta > s1_typing.tsv
```

And again, here is the same command with different formating for better readability (in scripts).

```bash
# typing with abricate
abricate --datadir /home/mf1/nambi24/flu-workflow/catalogues/INSAFLU\
    --db db_influenza_typing_v8 \
    -minid 70 -mincov 60 \
    s1_assembly.fasta \
    > s1_typing.tsv

```

